🐛 Fix: 채팅방 자동 퇴장 및 읽음 처리 기능 추가 보완

BE12-OZ opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: 
- 멤버 퇴장/추방 알림: 스터디 그룹에서 멤버가 나가거나 추방되었을 때, 이를 채팅방에 실시간으로 반영
- 자동 읽음 처리: 사용자가 채팅방에 입장하면 해당 방의 모든 메시지를 자동으로 읽음 처리

## 📄 상세 내용
- [x] 채팅방 멤버 퇴장/추방 시 실시간 알림 기능 추가 보완
- studies와 같은 다른 앱에서 호출하여 사용할 수 있도록 chat 앱에 서비스 함수를 구현
- apps/chat/services/chat_service.py :
   ChatRoomService에 broadcast_member_removal 메서드를 추가
- 동작:
   강제 접속 종료: 퇴장/추방된 사용자에게는 웹소켓 연결을 끊으라는 force_disconnect 이벤트를 전송
   시스템 메시지 전송: 채팅방에 남아있는 다른 멤버들에게는 "OO님이 스터디를 떠났습니다."와 같은 system_message를 전송
- apps/chat/consumers.py :
   위 두 이벤트(force_disconnect, system_message)를 수신하여 실제 동작을 수행하는 핸들러를 구현

- [x] 채팅방 입장 시 자동 읽음 처리 기능 추가 (프론트엔드 요청사항)
- 사용자가 채팅방에 들어왔을 때, 안 읽은 메시지 수를 갱신하기 위한 로직을 구현
- apps/chat/consumers.py :
   사용자가 웹소켓에 연결(connect)하면, mark_messages_as_read 라는 새로운 메서드가 호출됩니다.
   동작:
   해당 채팅방의 가장 최신 메시지를 조회 -> LastReadMessage 테이블에서 "현재 사용자가 이 채팅방의 몇 번 메시지까지 읽었는지"에 대한 정보를 최신 메시지로 업데이트
   이를 통해 채팅방에 입장하는 즉시 모든 메시지가 읽음 처리

- [x] Mypy 타입 오류 수정: ChatConsumer에서 인증된 사용자의 id 속성을 참조할 때 발생하던 타입 불일치 문제를 assert isinstance(user, User) 구문을 추가하여 명시적으로 타입을 검증함으로써 해결

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항
- [x] 프론트엔드 요청사항:
       확인/보완 필요 : '현재 접속 중인 인원'을 표시하는 기능 미구현 -> 체크 후 진행 예정 (1순위)
       
- [x] 신규 요청 내용 : 전체 안 읽은 메시지 수 총합 API
       요청 내용 : 모든 채팅방의 '안 읽은 메시지 수'를 합산하여 총합을 단일 API로 제공해주길 원함
       현재 상태 : 미구현 -> 구현 예정 (2순위)

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

BE12-OZ committed:
✨ feat: 채팅방 자동 퇴장 및 읽음 처리 기능 구현

chat 앱에 두 가지 실시간 기능을 추가하여 사용자 경험을 개선합니다.

1. 멤버 퇴장/추방 시 실시간 알림 기능
   - ChatRoomService에 다른 앱에서 호출 가능한  메서드 추가
   - 퇴장/추방된 유저에게  이벤트를, 방에 남은 유저에게  이벤트를 전송

2. 채팅방 입장 시 자동 읽음 처리 기능
   - ChatConsumer의  시점에 를 호출하여 를 최신 메시지로 업데이트

Mypy 타입 오류 수정
   - ChatConsumer에서 를 사용하여  관련 타입 오류 해결

Meoyoug reviewed (changes requested):
해당 기능에 대한 테스트코드가 없는것같아요

BE12-OZ commented:
체크해서 추가 보완하겠습니다. 감사합니다 !

BE12-OZ committed:
test(chat): Add unit test for mark_messages_as_read consumer logic

최종 요약

1. 문제점: mark_messages_as_read 기능에 대한 테스트 코드가 없다는 피드백을 받음.
2. 1차 시도 (통합 테스트): WebsocketCommunicator를 사용한 통합 테스트 추가 시, 테스트 환경의 비동기/트랜잭션 문제로 DoesNotExist 오류가 발생.
3. 2차 시도 (근본적 해결): TestCase로 기반을 변경해보았으나, connection already closed 오류가 발생하며 더 큰 문제가 발생.
4. 최종 해결 (단위 테스트): 변경 범위를 최소화하는 것으로 방향을 전환, 문제가 되는 connect 과정을 테스트하는 대신 mark_messages_as_read 함수의 로직만 검증하는 단위 테스트를 추가.
5. 마지막 오류 수정: 단위 테스트 중 발생한 SynchronousOnlyOperation 오류는 select_related를 사용하여 비동기 함수 내 동기 쿼리 문제를 해결.
